documentation/Using-the-SDK/Advanced/Creating-Your-Own-DataSource.md

```diff
@@ -0,0 +1,216 @@
+    // can take any other parameters that we like to help denote data.
+    public WebServiceDataSourceAttribute(
+        string serviceUri)
+        : base(typeof(WebServiceDataSource))
+    {
+        this.ServiceUri = new Uri(serviceUri);
+    }
+
+    public Uri ServiceUri { get; }
+}
+````
+
+## Accepts
+
+Finally, by default, all `DataSource`s of a given type will be routed to 
+`CustomDataSource`s that are decorated with the corresponding attribute. Currently,
+_ANY_ instance of a `WebServiceDataSource` will be routed to our `CustomDataSource`
+decorated with the `WebServiceDataSourceAttribute`, and the `IsDataSourceSupportedCore`
+method would have to do interrogation of said `DataSource` to determine if it is supported.
+It would be nice if to say "I only care about certain servces; don't even bother
+with others." This is what the `Accepts` method allows. The `Accepts` method is used
+to filter incoming `DataSource` objects before they ever reach your `CustomDataSource`
+so that your `IsDataSourceSupportedCore` logic is simpler. For example, the `FileDataSourceAttribute`
+in the SDK uses this method to reject anything that does not match the given
+file extension, so `CustomDataSource`s decorated with `FileDataSource(".txt")`, for
+example, only ever see `.txt` files. You should still implement
+`IsDataSourceSupportedCore` in your `CustomDataSource` to do the final determination.
+
+````cs
+// `DataSourceAttribute` has `AttributeUsage`, so you do not need to add that
+// attribute youself unless you need to override `AllowMultiple` or 'Inherited.`
+public sealed class WebServiceDataSourceAttribute
+    : DataSourceAttribute
+{
+    // our constructor will pass the DataSource Type to the base class. We
+    // can take any other parameters that we like to help denote our data. In
+    // this case, we simply say that data on the given server is supported.
+    public WebServiceDataSourceAttribute(
+        string serviceUri)
+        : base(typeof(WebServiceDataSource))
+    {
+        this.ServiceUri = new Uri(serviceUri);
+    }
+
+    public Uri ServiceUri { get; }
+
+    // implement this method to eagerly filter out data sources that are not
+    // accepted by the decorated Custom `DataSource`.
+    public override bool Accepts(IDataSource dataSource)
+    {
+        // IDataSource will always be of the Type based to the base
+        // class in our constructor.
+        var web = dataSource as WebServiceDataSource;
+
+        // This check is for defensive programming in case dataSource is
+        // ever null.
+        if (web is null)
+        {
+            return false;
+        }
+
+        // We only accept WebServiceDataSources that target our declared service.
+        return this.ServiceUri == web.ServiceUri;
+    } 
+}
+````
+
+# Using Your `DataSource` 
+
+Now that we have our new `DataSource`, we will see how to use our new `DataSource`
+with the Engine.
```

Comment:
this is basically the first time our docs mention the engine. Can we say something like "with the rest of the SDK?"